### PR TITLE
Moonbeam Watcher

### PR DESCRIPTION
**Pull request by bruce-riley**
Change-Id: Ic2c8b15438162fd097e866f9a0ca1f6adf6a3f55

---
This change is [Reviewable](https://reviewable.io/reviews/certusone/wormhole/1203)

### CODE REVIEW

File: node/pkg/celo/celoimpl.go [hunk 1]
```diff
@@ -156,14 +157,17 @@ func (e *CeloImpl) SubscribeNewHead(ctx context.Context, sink chan<- *ethTypes.H
...
 
-	// The purpose of this is to map events from the Celo event channel to the Eth event channel.
+	// The purpose of this is to map events from the Celo event channel to the new block event channel.
 	go func() {
 		for {
 			select {
 			case <-ctx.Done():
 				return
-			case celoEvent := <-headSink:
-				sink <- convertNewHeadEventToEth(celoEvent)
```
Comment (evan-gray) [inline node/pkg/celo/celoimpl.go:166]: We don't need this function down below any more then 🙂

File: node/pkg/ethereum/moonbeamimpl.go [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+	latestBlockNumber, err := e.getLatestBlockNumber(ctx)
+	if err != nil {
+		return nil, err
+	}
+	currentBlockNumber := latestBlockNumber
+
+	const DELAY_IN_MS = 1000
+
+	timer := time.NewTimer(1 * time.Millisecond) // Start immediately.
```
Comment (evan-gray) [inline node/pkg/ethereum/moonbeamimpl.go:134]: can omit the `1 *` here

File: node/pkg/ethereum/moonbeamimpl.go [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+						if currentBlockNumber > latestBlockNumber {
+							e.logger.Info("waiting for new latest block", zap.String("eth_network", e.BaseEth.NetworkName),
+								zap.Uint64("current_block", currentBlockNumber), zap.Uint64("latest_block", latestBlockNumber))
+							break
+						}
+					}
+
+					// Fetch the hash every time, in case it changes due to a rollback.
+					blockHash, err := e.getBlockHash(ctx, currentBlockNumber)
```
Comment (evan-gray) [inline node/pkg/ethereum/moonbeamimpl.go:159]: this could return a NewBlock

File: node/pkg/ethereum/moonbeamimpl.go [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+						break
+					}
+					if !finalized {
+						break
+					}
+
+					e.logger.Info("block is now finalized", zap.String("eth_network", e.BaseEth.NetworkName),
+						zap.Uint64("block", currentBlockNumber), zap.Stringer("hash", blockHash))
+
+					ev, err := e.getBlock(ctx, currentBlockNumber)
```
Comment (evan-gray) [inline node/pkg/ethereum/moonbeamimpl.go:181]: If the `getBlockHash` call above instead returned a NewBlock, you could avoid this call.

File: node/pkg/ethereum/moonbeamimpl.go (hunk 3 above)
Comment (evan-gray) [inline node/pkg/ethereum/moonbeamimpl.go:159]: I feel like there's some small inefficiency between this and the `getLatestBlockNumber` call above, though only the first loop when currentBlockNumber == latestBlockNumber. Not immediately sure how to make it more efficient.

File: node/pkg/ethereum/moonbeamimpl.go [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+	err := e.mbRpcCon.CallContext(ctx, &m, "eth_getBlockByNumber", "latest", false)
+	if err != nil {
+		e.logger.Error("failed to get current block", zap.String("eth_network", e.BaseEth.NetworkName), zap.Error(err))
+		return 0, err
+	}
+
+	return strconv.ParseUint(m.Number.String()[2:], 16, 64)
+}
+
+func (e *MoonbeamImpl) getBlockHash(ctx context.Context, number uint64) (ethCommon.Hash, error) {
```
Comment (evan-gray) [inline node/pkg/ethereum/moonbeamimpl.go:223]: I think there's a way to combine this with `getLatestBlockNumber`, check out https://github.com/ethereum/go-ethereum/blob/440c9fcf75d9d5383b72646a65d5e21fa7ab6a26/ethclient/ethclient.go#L533-L542. Also should number be `*big.Int`?

File: node/pkg/ethereum/moonbeamimpl.go [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+	type Marshaller struct {
+		Hash ethCommon.Hash `json:"hash"`
+	}
+
+	var m Marshaller
+	err := e.mbRpcCon.CallContext(ctx, &m, "eth_getBlockByNumber", strconv.FormatUint(number, 10), false)
```
Comment (evan-gray) [inline node/pkg/ethereum/moonbeamimpl.go:229]: you won't need the string conversion with the helper linked above

File: node/pkg/ethereum/moonbeamimpl.go [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+	}
+
+	var m Marshaller
+	err := e.mbRpcCon.CallContext(ctx, &m, "eth_getBlockByNumber", "latest", false)
+	if err != nil {
+		e.logger.Error("failed to get current block", zap.String("eth_network", e.BaseEth.NetworkName), zap.Error(err))
+		return 0, err
+	}
+
+	return strconv.ParseUint(m.Number.String()[2:], 16, 64)
```
Comment (evan-gray) [inline node/pkg/ethereum/moonbeamimpl.go:220]: This should probably just return m.Number

**Review by evan-gray — CHANGES_REQUESTED** [review]
This approach is looking really promising!

File: node/pkg/ethereum/moonbeamimpl.go [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+	common "github.com/certusone/wormhole/node/pkg/common"
+	ethAbi "github.com/certusone/wormhole/node/pkg/ethereum/abi"
+
+	"go.uber.org/zap"
+)
+
+type MoonbeamImpl struct {
+	BaseEth  *EthImpl
+	logger   *zap.Logger
+	mbRpcCon *ethRpc.Client
```
Comment (SEJeff) [inline node/pkg/ethereum/moonbeamimpl.go:28]: I understand what this is, but `client` seems a bit nicer wrt naming.

File: node/pkg/ethereum/moonbeamimpl.go [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+					}
+					if !finalized {
+						break
+					}
+
+					sink <- block
+					currentBlockNumber.Add(&currentBlockNumber, BIG_ONE)
+				}
+
+				timer = time.NewTimer(DELAY_IN_MS * time.Millisecond)
```
Comment (SEJeff) [inline node/pkg/ethereum/moonbeamimpl.go:177]: Using a variable for that instead of a magic number. 👍

File: node/pkg/ethereum/moonbeamimpl.go [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+		return nil, err
+	}
+
+	n := big.Int(*m.Number)
+	return &common.NewBlock{
+		Number: &n,
+		Hash:   m.Hash,
+	}, nil
+}
+
```
Comment (SEJeff) [inline node/pkg/ethereum/moonbeamimpl.go:216]: Might be nice to add a comment somewhere about why `isBlockFinalized` is even necessary. I know why now, but someone new might not understand why.

**Review by SEJeff — APPROVED** [review]

File: node/pkg/ethereum/moonbeamimpl.go [hunk 11]
```diff
@@ -103,13 +104,22 @@ func (sub *MoonbeamSubscription) Unsubscribe() {
 	})
 }
 
+// Moonbeam can publish blocks before they are marked final. This means we need to sit on the block until a special "is finalized"
+// query returns true. The assumption is that every block number will eventually be published and finalized, it's just that the contents
+// of the block (and therefore the hash) might change if there is a rollback. Therefore rather than subscribing for headers from geth,
+// we use a polling mechanism to get the next expected block, and keep doing it until it is marked final.
```
Comment (SEJeff) [inline node/pkg/ethereum/moonbeamimpl.go:110]: 👍   this is excellent. Thankyou!

**Review by SEJeff — APPROVED** [review]

**Review by SEJeff — APPROVED** [review]

**Review by SEJeff — APPROVED** [review]

**Pull request merged**